src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1417,6 +1427,17 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     return loweredBody;
                 }
 
+                if (method.ContainingType.TryGetCorrespondingStaticMetadataExtensionMember(method) is MethodSymbol metadataMethod)
+                {
+                    var extensionRewriter = new InstanceExtensionMethodBodyRewriter(method, (SourceExtensionMetadataMethodSymbol)metadataMethod);
+                    loweredBody = (BoundStatement)extensionRewriter.Visit(loweredBody);
+                    method = metadataMethod;
+                }
+                else
+                {
+                    loweredBody = InstanceExtensionMethodReferenceRewriter.Rewrite(method, loweredBody);
```

Comment:
The body of an instance extension method may reference instance methods from other extension types so shouldn't it also get the `InstanceExtensionMethodReferenceRewriter` treatment?

Reply:
> The body of an instance extension method may reference instance methods from other extension types so shouldn't it also get the `InstanceExtensionMethodReferenceRewriter` treatment?

It does, `InstanceExtensionMethodBodyRewriter` includes the same logic. This way we save on an extra rewrite pass, but the logic is duplicated to some degree. Unfortunately there is no common base between the rewriters where the logic could belong, and there is no inheritance relationship between the two rewriters. There is, however, an advantage from doing the work during `InstanceExtensionMethodBodyRewriter` apart from saving on an extra pass. We are rewriting `BoundThisReference`s to the synthesized parameter references. Those, have different types. By doing call rewrite at the same time we never end up with `BoundCall` nodes having type mismatch between "this" value and the target synthesized parameter. Quite possibly this (the mismatch) is not a big deal, since extension instances can come from sources other than `this` references. I will add a PROTOTYPE comment to reevaluate this later.

Reply:
Would it be possible to only use `InstanceExtensionMethodBodyRewriter`, but disable/skip the parts that only apply to extension method bodies?

Reply:
At the moment ```InstanceExtensionMethodBodyRewriter``` is very specialized for the task of rewriting in a different context. Having two different rewriters feel simpler clearer to me.